feat(file sink): support Zstandard-compressed output

hdhoang opened the pull request:
This is part of #2302

I'm a bit uncomfortable with raising `zstd` to a true dependency, but the `lines_from_zstd_file` test function is always built. I put in the current version `0.10` from Cargo.lock.

todos:
- [x] bump to 0.11 or later, without default features
- [x] show zstd-commandline compatibility
- [x] perhaps gate this & `flate2` under dev-dependencies, since they are used for tests only?
- [x] explicitly say we don't special zstd feature like `dictionary`

zamazan4ik commented:
+1 for the PR. I am interested in enabling zstd support to the sinks. The use-case is simple: write compressed file and later work with directly via tooling like `zstdgrep` or decompressing and then working with usual tooling.

@hdhoang Just in case: did you try decompress resulting files after Vector with `zstd` command line tool? Does it work?

zamazan4ik commented:
@jszwedko just a friendly ping to review it :)

zamazan4ik commented:
@hdhoang by the way, I guess we need somehow handle compression levels (as you mentioned before - https://github.com/vectordotdev/vector/pull/3032). As far as I see, this PR does not support adding custom compression levels.

jszwedko reviewed (commented):
Thanks for this contribution @hdhoang ! Just one question below.

bruceg commented on Cargo.toml:
```diff
@@ -303,6 +303,7 @@ typetag = { version = "0.2.3", default-features = false }
 url = { version = "2.2.2", default-features = false, features = ["serde"] }
 uuid = { version = "1", default-features = false, features = ["serde", "v4"] }
 warp = { version = "0.3.1", default-features = false }
+zstd = "0.10.0"
```
Would it also make sense to make this optional, behind a feature flag?

jszwedko replied:
I thought about that too, but we already have `flate2` in here as a general dependency. We could make it optional and only include it for the `file` source.

neuronull replied:
This is in the initial comment (as to why `0.10`) :

> I put in the current version 0.10 from Cargo.lock

Sounds like the latest version should be used though.

jszwedko replied:
Aha, yes, I missed that. Leaving it at 0.10 is fine for the purposes of this PR. We can update it separately.

hdhoang replied:
i have moved `zstd` under dev-dep, and disabled default-features. It's only used in 1 test, while flate2 is used more generally

neuronull commented on src/sinks/file/mod.rs:
```diff
@@ -96,6 +96,9 @@ pub enum Compression {
     /// Gzip compression.
     Gzip,
 
+    /// Zstandard compression.
```
The new compression algorithm should be added to `website/cue/reference/components/sinks/file.cue`

hdhoang replied:
done in the thread below, but we have question about levels

zamazan4ik commented on website/cue/reference/components/sinks/file.cue:
```diff
@@ -20,7 +20,7 @@ components: sinks: file: {
 				default: "none"
-				algorithms: ["none", "gzip"]
+				algorithms: ["none", "gzip", "zstd"]
 				levels: ["none", "fast", "default", "best", 0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
```
By the way, could someone explain me, where in the Vector documentation these compression levels are described? I cannot find related information [here](https://vector.dev/docs/reference/configuration/sinks/file/#compression). Is it rendered at all?

jszwedko replied:
Yeah, it looks like that isn't currently being rendered correctly on the website. The `file` source also doesn't appear to actually allow for configuration of `levels` (though other places `compression` is used does).

zamazan4ik replied:
So should we do with a compression level? Fix them in a separate PR?

hdhoang replied:
i suppose so, both for rendering & how to match them with available algos

jszwedko replied:
I think we can just drop `levels` here since it isn't actually used by the `file` sink. When we add support for compression levels there, we can re-add it.

zamazan4ik replied:
So in this case I guess we need to at least specify, which default level is used for each compression algorithm (since Vector uses some defaults). It could be kinda surprising for the users, if the default level is 0 (no or almost no compression).

jszwedko replied:
Yeah, that's fair. You can update the algorithm descriptions here:

https://github.com/vectordotdev/vector/blob/553763f6ed37f1bfda7655f9aa6679e2423d27f9/website/cue/reference/components/sinks.cue#L146-L162

To include their default levels.

hdhoang replied:
ah right, I added a generic note about Default level being used, but i will add numeric equivalent for each algo too. i will then check the cue whitespace issue

hdhoang replied:
oops, the doc here turns out important for zstd too. [flate2 Default is `6`](https://docs.rs/flate2/1.0.24/src/flate2/lib.rs.html#194) which is sensible.

However, async_compression set [zstd `Default` as `0`](https://docs.rs/async-compression/0.3.12/src/async_compression/lib.rs.html#247), rather than the [libzstd value of `3`](https://docs.rs/zstd/latest/zstd/constant.DEFAULT_COMPRESSION_LEVEL.html).

I think we can suggest `async-compression` to use public constant, and then update vector documentation later.

zamazan4ik replied:
I suggest document the current default value right now. When `async-compression` will update their default compression level, we will update Vector documentation too.

Reasoning is simple: we do not know, when `async-compression` will update their code.

zamazan4ik commented:
Another question from me. What is the current behavior in the following scenario:

1. Set up File sink with Gzip compression
2. Start logging something
3. Change Gzip to Zstd in the config for the File sink and reload Vector
4. Continue logging something

Am I right that in this case we will got different compression algorithms for the same file and this file would be semantically corrupted? Is it desired behaviour or we need somehow mitigate it?

jszwedko commented:
> Another question from me. What is the current behavior in the following scenario:
> 
> 1. Set up File sink with Gzip compression
> 2. Start logging something
> 3. Change Gzip to Zstd in the config for the File sink and reload Vector
> 4. Continue logging something
> 
> Am I right that in this case we will got different compression algorithms for the same file and this file would be semantically corrupted? Is it desired behaviour or we need somehow mitigate it?

I feel like this is an edge case we can just document as a warning around the `compression` option for the `file` sink. If the compression is changed, I imagine users will also be changing the filename to use the new correct extension based on the compression being used. You are correct though, if it appends to an existing file, the file will be corrupt. I'd welcome a mitigation for it, I'm just not sure it is worth the effort given how unlikely it is users will hit it.

zamazan4ik commented:
@hdhoang do you need a help to work on this feature? I think I can try to help on it if you have a lack of time.

hdhoang commented:
thanks! i still work on it these few days.

>  mitigation for it

i added a note about compression and file extension to `path` cue doc, since current example says `.log`. i can add more examples too

hdhoang commented:
i dropped the `levels` line in `file.cue`, but validator wanted it back:

```text
Validating cue files correctness...
Failed!

components.sinks.file.features.send.compression.levels.0: incomplete value "none" | "fast" | "default" | "best" | >=0 & <=9
```

Because i added the "unless otherwise specified" phrasing to `sinks.cue`, and there's no way to specify, I'll return the `levels` line.

hdhoang commented:
hi, i think i have addressed your comments. let's have another round of discussion!

i'm building a dev binary now, and would try out some scenarios like partial kill & check that zst files are decompressible even with broken half-stream inside.

---

i have some othe comments about the dev setup with `make environment`.

1. the rustup part isn't cached in a docker volume, so each time i enter environment (eg after a build OOM), it has to fetch rustc again. Perhaps we can request a small volume to cache this?
2. the `git` command inside environment doesn't want to run out-of-the-box because of different UID. Perhaps the image can set `safe.directory` unconditionally?

```sh
root@jer:/git/vectordotdev/vector# git status
fatal: unsafe repository ('/git/vectordotdev/vector' is owned by someone else)
To add an exception for this directory, call:

        git config --global --add safe.directory /git/vectordotdev/vector
```

3. i tried to use podman at the start, but got nowhere because the command want to pass dockerd socket into the environment. With docker i can do fine though.

Sorry about the long digression, if any point is valid, would you accept some more docs fix? Cheers!

jszwedko commented:
> i have some othe comments about the dev setup with `make environment`.

These are great, thanks for the suggestions!

> 1. the rustup part isn't cached in a docker volume, so each time i enter environment (eg after a build OOM), it has to fetch rustc again. Perhaps we can request a small volume to cache this?

Yeah, agreed, it seems like we could use volumes more to cache.

> 2. the `git` command inside environment doesn't want to run out-of-the-box because of different UID. Perhaps the image can set `safe.directory` unconditionally?
> 
> ```shell
> root@jer:/git/vectordotdev/vector# git status
> fatal: unsafe repository ('/git/vectordotdev/vector' is owned by someone else)
> To add an exception for this directory, call:
> 
>         git config --global --add safe.directory /git/vectordotdev/vector
> ```

Agreed. I think people usually end up commiting from the outside, but that seems like a sensible change to me.

> 3. i tried to use podman at the start, but got nowhere because the command want to pass dockerd socket into the environment. With docker i can do fine though.

Aha, interesting, yeah. Open to PRs improving podman support too.

> Sorry about the long digression, if any point is valid, would you accept some more docs fix? Cheers!

Definitely! I think we can split off any of these into a separate PR, if you agree.

---

It sounds like you are still doing some testing, but just let us know whenever you think this is good to go!

hdhoang commented:
i think this is good to go now. thanks everyone for the reviews!

zamazan4ik commented:
@jszwedko I guess this PR could be merged.
